src/vmm/src/devices/virtio/queue.rs

```diff
@@ -522,11 +522,10 @@ impl Queue {
     pub fn is_valid<M: GuestMemory>(&self, mem: &M) -> bool {
         if !self.is_layout_valid(mem) {
             false
-        } else if self.len(mem) > self.max_size {
+        } else if self.size > self.max_size {
```

Comment:
`self.size <= self.max_size` is already checked in `is_layout_valid`. But this is also a fundamentally different check. Checking `size` against `max_size` checks whether the queue configuration the guest passed us via MMIO before activation makes sense at all. Checking that the `len` does not exceed `max_size` (well, I guess really this should be `size`?) is a runtime check that ensures the guest is using the queue correctly, e.g. it doesn't tell us there are more available descriptors than is mathematically possible. I'm not sure why you want to drop that check

Reply:
The reason why I need to remove the `len(mem)`  from here is that the validation of queue happen before device activation. So with new changes the `len` method will use null pointers to try to calculate len, so it will fail.
Changing the logic of testing if a queue is valid seems beyond this PR, even though it would totally make sense to do the validation check in `activate` method and not before it. Also the `is_valid` is not be called during an actual runtime because all `is_valid` checks are in the `debug_assert!` macros, so at runtime guest can easily "make up" queue len bigger than it should be.
I see now that `is_layout_valid` is indeed check the `self.size <= self.max_size`, so I can remove the `is_valid` all together.

Reply:
Mh, that makes sense, thanks for the explanation! I think the `is_valid` thing used to also do the check after snapshot restore, but that's of questionable value, too, I guess, because `pop` already has this "guest doesn't set avail_idx to nonsense" check.

Reply:
If we do this check on `pop` each time, do we really need to do it one more time on activation/initialization? I will adjust the commit message about this. Also as far as I can think, even if guest tells us to pop more descriptors than the queue should hold, we will just go around the queue and start poping same valid descriptors again.

Reply:
but if you strongly want this check to be in `initialize`, I can add it back.

Reply:
added it back as an assert, so it will have same behavior as `pop`